File: java/leetcode/src/main/java/com/leetcode/BFS_DFS/Subsets_78.java
package com.leetcode.BFS_DFS;

import java.util.ArrayList;
import java.util.List;

public class Subsets_78 {
    List<List<Integer>> sets = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        List<Integer> set = new ArrayList<>();
        sets.add(new ArrayList<>());
        dfs(nums, 0, set);


        return sets;
    }

    void dfs(int[] nums, int start, List<Integer> set) {
        for(int i = start; i < nums.length; i++) {
            set.add(nums[i]);
            sets.add(new ArrayList<>(set));
            dfs(nums, i + 1, set);
            set.remove(set.size() - 1);
        }
    }

    public  static void main(String[] args) {
        int[] nums = new int[]{1, 2};
        Subsets_78 test = new Subsets_78();
        List<List<Integer>> res = test.subsets(nums);

        for (List<Integer> l : res) {
            System.out.println(l.toString());
        }
    }
}

File: js/algrithm/huahua130/bfs-dfs/Subsets_78.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const ans = []
    const dfs = function(start, path) {
        ans.push(path.slice())
        for (let i = start; i < nums.length; i++) {
            path.push(nums[i])
            dfs(i + 1, path)
            path.pop()
        }
    }
    dfs(0, [])
    return ans
};